model/task_lifecycle.go

```diff
@@ -775,48 +775,104 @@ func RestartItemsAfterVersion(cq *commitqueue.CommitQueue, project, version, cal
 	return catcher.Resolve()
 }
 
-// DequeueAndRestartForTask restarts all items after the given task's version, aborts/dequeues the current version,
-// and sends an updated status to GitHub.
+// DequeueAndRestartForTask restarts all items after the given task's version,
+// aborts/dequeues the current version, and sends an updated status to GitHub.
 func DequeueAndRestartForTask(cq *commitqueue.CommitQueue, t *task.Task, githubState message.GithubState, caller, reason string) error {
-	if cq == nil {
-		var err error
-		cq, err = commitqueue.FindOneId(t.Project)
+	mergeErrMsg := fmt.Sprintf("commit queue item '%s' is being dequeued: %s", t.Version, reason)
+	if t.Details.Type == evergreen.CommandTypeSetup {
+		// If the commit queue merge task failed on setup, there is likely a merge conflict.
+		mergeErrMsg = "Merge task failed on setup, which likely means a merge conflict was introduced. Please try merging with the base branch."
+	}
+
+	_, err := dequeueAndRestartItem(dequeueAndRestartOptions{
+		cq:           cq,
+		projectID:    t.Project,
+		versionID:    t.Version,
+		taskID:       t.Id,
+		caller:       caller,
+		reason:       reason,
+		mergeErrMsg:  mergeErrMsg,
+		githubStatus: githubState,
+	})
+	return err
+}
+
+// DequeueAndRestartForVersion restarts all items after the commit queue
+// item, aborts/dequeues this version, and sends an updated status to GitHub. If
+// it succeeds, it returns the removed item.
+func DequeueAndRestartForVersion(cq *commitqueue.CommitQueue, project, version, user, reason string) (*commitqueue.CommitQueueItem, error) {
```

Comment:
Can we consolidate this with `DequeueAndRestartForTask` to a single `DequeueAndRestartItem` function we can refer to for this behavior? Since it seems they do the same thing mod a little bit of task / error message handling if a task has been passed in.

Reply:
I usually prefer making two different functions and calling a helper when it simplifies the complexity of reading it. My experience in the Evergreen code base is that too much functionality gets merged and consolidated to reduce the number of function definitions, but at the expense of increasing complexity substantially (and therefore producing misunderstandings and bugs). If possible, I prefer to lean towards obvious and easy-to-use function calls, as opposed to a few function calls that are complicated and you really have to dig to understand how to call it properly.

I did try consolidating the two initially, but they're annoyingly just a bit different enough that merging them into one function would just be more confusing. In the case where you're dequeueing an entire version, you would have to pass a nil task and know that it's safe to pass a nil task, and understand why you would pass a task vs. just a version. You could alternatively push the task-specific logic into the callers, but that doesn't seem any simpler than having two functions. Splitting them in two makes it obvious that the task one is handling dequeue after a commit queue task fails, whereas the other one is for removing entire versions for reasons not realted to tasks. This reduces overhead of understanding what conceptually is being done in the two different cases.

Reply:
Alright, that's a fair point, I'm on board. Thanks for the thorough response